Lib/extractor/formats/opentype.py

```diff
@@ -1078,3 +1082,110 @@ def extractOpenTypeFeatures(source):
     if _haveFontFeatures:
         return unparse(source).asFea()
     return ""
+
+
+# -------
+# Anchors
+# -------
+
+
+def extractAnchors(source, destination):
+    gpos = source["GPOS"].table
```

Comment:
You need to check if the source font does have a "GPOS" table at all.

Reply:
Anything else I should change?

Reply:
I've added the check.

Reply:
Oh, cool, the tests now pass!

However, it looks like these roundtrip tests currently only check that a font that has been decompiled can be compiled back successfully, but they don't seem to check that specific tables/features are written into the output UFO. Perhaps they should?

Reply:
Yeah, the tests are severly underdeveloped in this module :-/

Reply:
To be fair, extractor is not ever meant to do the impossible task of a perfect 1:1 extraction of binary to UFO source (as in, one will never get things back like anchor names, etc)

Reply:
> To be fair, extractor is not ever meant to do the impossible task of a perfect 1:1 extraction of binary to UFO source (as in, one will never get things back like anchor names, etc)

Of course. It should do as much as it can though, right? If anchor names are never exported into the binary font file, and hence can't be extracted, that's absolutely fine. As you can see, I just call them `Anchor-1`, `Achor-2` and so on. It would be nice, however, to be able to tell with confidence that a font decompiled using extractor can be recompiled again with no losses, or to at least have clear knowledge of what will be preserved and/or what will be lost during a roundtrip.

As an aside, I'm currently in the process of converting a proprietary font family into an open-source one, and, having very little prior experience with fonts, I just want to make sure that I don't lose anything of importance and relevance in the process. The fonts were last updated over ten years ago and of course, no VCS was ever involved. Even though I received the work files, I figured it's probably more reliable for me to actually use the last version TTF files as the source for conversion instead of all these FontLab, VOLT and other files, of which who knows which ones were latest/final. It seems to have worked quite well so far, although I'll admit `fontmake` (which I'm currently using for compilation) makes some opinionated decisions which make it difficult for me to figure out what exactly is being lost and how useful it is. Also, I don't understand the hinting bytecode at all, so I'm not sure If I (or anyone) will ever be able to change it meaningfully should there be a need for that (and that would be definitely useful with reversed contours, for example).

Reply:
You could convert the FontLab files (assuming they are FontLab 5's VFB format) with our [vfbLib](https://pypi.org/project/vfbLib/) to UFO and compare with the UFOs produced by extractor to get an idea what data may be available or missing.

Reply:
> You could convert the FontLab files (assuming they are FontLab 5's VFB format) with our [vfbLib](https://pypi.org/project/vfbLib/) to UFO and compare with the UFOs produced by extractor to get an idea what data may be available or missing.

Hmmm... The resulting UFOs are substantially different. Looking at just one glyph (`A`), it looks like there are very few matching lines between the two files: extractor extracts advance, unicode, anchors (after my patch), outline contour points without names, and TrueType instructions, meanwhile `vfb3ufo` exports advance, unicode, guidelines, points with names, PostScript hints (`com.adobe.type.autohint.v2`), and FontLab ttprogram (`com.fontlab.ttprogram`). Ouch!

Reply:
To be fair, both tools have somewhat different goals. `vfbLib` was made to get all necessary data out of VFB files that we (@LucasFonts) need to build variable fonts from UFO + designspace.

Extractor originates from a time where anchors were not widely used in (Latin) font production, so it concentrated on the parts that were most important to recreate source files to keep working on the designs, i.e. outlines and kerning.

Of course, `vfb3ufo` can put stuff into the UFO that simply doesn't exist in a binary font, like guidelines, point labels, anchor names.

The FontLab ttprogram is a high-level representation of its TrueType hinting, and can't easily be compiled into TrueType assembly or bytecode again, so in that regard extractor will give you better results.

Reply:
@jenskutilek, vfb3ufo actually doesn't seem to export anchors at all. I've filed a feature request at https://github.com/LucasFonts/vfbLib/issues/104 for you. :)